File: typescript/src/02-parameterize-constructor/discount-calculator.ts
import { SystemClock } from "./system-clock.js";
import { orderSubtotalCents, type Cents, type Coupon, type Order } from "../domain.js";

export interface PricedOrder {
  subtotalCents: Cents;
  discountCents: Cents;
  totalCents: Cents;
  /** Why the coupon was not applied, if it was rejected. */
  rejectedReason?: string;
}

/**
 * Applies a coupon to an order, honouring the coupon's expiry.
 *
 * The behaviour we want to test is time-dependent: an expired coupon must be
 * rejected and a still-valid one must be applied. But `this.clock` is built
 * with `new SystemClock()` right here in the constructor, so the calculator is
 * hard-wired to "now". A test cannot pin the clock to a known instant, which
 * makes the expiry branch effectively untestable.
 */
export class DiscountCalculator {
  private readonly clock = new SystemClock();

  price(order: Order, coupon: Coupon | undefined): PricedOrder {
    const subtotalCents = orderSubtotalCents(order);

    if (!coupon) {
      return { subtotalCents, discountCents: 0, totalCents: subtotalCents };
    }

    if (coupon.expiresAt.getTime() <= this.clock.now().getTime()) {
      return {
        subtotalCents,
        discountCents: 0,
        totalCents: subtotalCents,
        rejectedReason: "coupon expired",
      };
    }

    const discountCents = Math.round((subtotalCents * coupon.percentOff) / 100);
    return {
      subtotalCents,
      discountCents,
      totalCents: subtotalCents - discountCents,
    };
  }
}

File: java/src/main/java/com/qmates/kata/parameterizeconstructor/DiscountCalculator.java
package com.qmates.kata.parameterizeconstructor;

import com.qmates.kata.domain.Coupon;
import com.qmates.kata.domain.Order;
import com.qmates.kata.domain.Orders;

/**
 * Applies a coupon to an order, honouring the coupon's expiry.
 *
 * <p>The behaviour we want to test is time-dependent: an expired coupon must be
 * rejected and a still-valid one must be applied. But {@code clock} is built
 * with {@code new SystemClock()} right here in the field initializer, so the
 * calculator is hard-wired to "now". A test cannot pin the clock to a known
 * instant, which makes the expiry branch effectively untestable.
 */
public class DiscountCalculator {

  private final SystemClock clock = new SystemClock();

  public PricedOrder price(Order order, Coupon coupon) {
    long subtotal = Orders.subtotalCents(order);

    if (coupon == null) {
      return PricedOrder.of(subtotal, 0);
    }

    if (!coupon.expiresAt().isAfter(clock.now())) {
      return PricedOrder.rejected(subtotal, "coupon expired");
    }

    long discount = Math.round(subtotal * coupon.percentOff() / 100.0);
    return PricedOrder.of(subtotal, discount);
  }
}
